objax/variable.py

```diff
@@ -53,15 +53,44 @@ def value(self) -> JaxArray:
     def value(self, tensor: JaxArray):
         raise NotImplementedError('Pure method')
 
-    def assign(self, tensor: JaxArray):
+    def assign(self, tensor: JaxArray, check=True):
         """Sets the value of the variable."""
+        if check:
+            self.assert_assigned_type_and_shape_match(tensor)
         self.value = tensor
 
     def reduce(self, tensors: JaxArray):
         """Method called by Parallel and Vectorize to reduce a multiple-device (or batched in case of vectoriaation)
         value to a single device."""
         if self._reduce:
-            self.assign(self._reduce(tensors))
+            self.assign(self._reduce(tensors), check=False)
+
+    def assert_assigned_type_and_shape_match(self, tensor):
+        assert isinstance(tensor, JaxArray.__args__), \
+            f'Assignments to variable must be an instance of JaxArray,  but received f{type(tensor)}.'
+
+        def shape_and_device(array):
```

Comment:
This function should be also defined in `objax.util.check`, defining a function inside a function may incur a runtime penalty.